### DISCUSSION TITLE
6 channel DMX controller with

### DISCUSSION THREAD

**Discussion by ghost** [dbuezas/lgt8fx#226] [Q&A] [answered]
I try to make a DMX controller with a LGT8F328P-32 (LQFP32 PRO MINI NANO3;
The controller is based on this project I found on instructables.com.
https://www.instructables.com/Build-Your-Own-DMX-Fixture-Arduino/

The only modifications I made are replaced the ATmega328 by the LGT8F328 board and switched from one channel to 6 channels.
I can compile and upload the sketch to the board.

[code]
#define ARDUINO_ARCH_AVR; // uart0 voor ardiuino mini, nano, uno
#include <DMXSerial.h>      // get the libary at:  https://github.com/mathertel/DMXSerial  by: mathertel 
#include <util/delay_basic.h>  // replace modified _delay_loop_2() for LGT8F328P

// Constants for the program
const int EnablePin1 = 3;  
const int EnablePin2 = 5;  
const int EnablePin3 = 6;  
const int EnablePin4 = 9;  
const int EnablePin5 = 10;  
const int EnablePin6 = 11;  

// inputs from the dip switches ( starting address)
const int Input1 = A7;  //
const int Input2 = A6;  //
const int Input3 = A5;  //
const int Input4 = A4;  //
const int Input5 = A0;  //
const int Input6 = A1; //
const int Input7 = A2; //
const int Input8 = A3; //

// container for the dip switch value
int Dip1 = 0;
int Dip2 = 0;
int Dip3 = 0;
int Dip4 = 0;
int Dip5 = 0;
int Dip6 = 0;
int Dip7 = 0;
int Dip8 = 0;

int startAddress = 1;  

void setup () {
 
  DMXSerial.init(DMXReceiver);
   
  pinMode(EnablePin1, OUTPUT);
  pinMode(EnablePin2, OUTPUT);
  pinMode(EnablePin3, OUTPUT);
  pinMode(EnablePin4, OUTPUT);
  pinMode(EnablePin5, OUTPUT);
  pinMode(EnablePin6, OUTPUT);
  
  pinMode(Input1, INPUT);
  pinMode(Input2, INPUT);
  pinMode(Input3, INPUT);
  pinMode(Input4, INPUT);
  pinMode(Input5, INPUT);
  pinMode(Input6, INPUT);
  pinMode(Input7, INPUT);
  pinMode(Input8, INPUT);
  
// test outputs at startup/reset DMXcontroller
  digitalWrite(EnablePin1, 0); // turn lights off
  digitalWrite(EnablePin2, 0); // turn lights off
  digitalWrite(EnablePin3, 0); // turn lights off
  digitalWrite(EnablePin4, 0); // turn lights off
  digitalWrite(EnablePin5, 0); // turn lights off
  digitalWrite(EnablePin6, 0); // turn lights off
delay (2000);
  digitalWrite(EnablePin1, 1); // turn lights on
  digitalWrite(EnablePin2,1); // turn lights on
  digitalWrite(EnablePin3,1); // turn lights on
  digitalWrite(EnablePin4,1); // turn lights on
  digitalWrite(EnablePin5,1); // turn lights on
  digitalWrite(EnablePin6,1); // turn lights on
delay (2000);
  digitalWrite(EnablePin1, 0); // turn lights off
  digitalWrite(EnablePin2, 0); // turn lights off
  digitalWrite(EnablePin3, 0); // turn lights off
  digitalWrite(EnablePin4, 0); // turn lights off
  digitalWrite(EnablePin5, 0); // turn lights off
  digitalWrite(EnablePin6, 0); // turn lights off
}

void loop() {
  // Calculate how long no data backet was received
  unsigned long lastPacket = DMXSerial.noDataSince();

  // translate the dip switches inputs into a number
 if (digitalRead(Input1) == HIGH) {Dip1 = 1;}
 if (digitalRead(Input1) == LOW)  {Dip1 = 0;}

 if (digitalRead(Input2) == HIGH) {Dip2 = 2;}
 if (digitalRead(Input2) == LOW)  {Dip2 = 0;}

 if (digitalRead(Input3) == HIGH) {Dip3 = 4;}
 if (digitalRead(Input3) == LOW)  {Dip3 = 0;}

 if (digitalRead(Input4) == HIGH) {Dip4 = 8;}
 if (digitalRead(Input4) == LOW)  {Dip4 = 0;}

 if (digitalRead(Input5) == HIGH) {Dip5 = 16;}
 if (digitalRead(Input5) == LOW)  {Dip5 = 0;}

 if (digitalRead(Input6) == HIGH) {Dip6 = 32;}
 if (digitalRead(Input6) == LOW)  {Dip6 = 0;}

 if (digitalRead(Input7) == HIGH) {Dip7 = 64;}
 if (digitalRead(Input7) == LOW)  {Dip7 = 0;}

 if (digitalRead(Input8) == HIGH) {Dip8 = 128;}
 if (digitalRead(Input8) == LOW)  {Dip8 = 0;}

 startAddress = Dip1 + Dip2 + Dip3 + Dip4 + Dip5 + Dip6 + Dip7 + Dip8; // set the startAddress 

      if (lastPacket < 2000) {                  // it is less than 1 seconds after we received a DMX value 
    analogWrite(EnablePin1, DMXSerial.read(startAddress));  
    analogWrite(EnablePin2, DMXSerial.read(startAddress +1));          
    analogWrite(EnablePin3, DMXSerial.read(startAddress +2));          
    analogWrite(EnablePin4, DMXSerial.read(startAddress +3));          
    analogWrite(EnablePin5, DMXSerial.read(startAddress +4));          
    analogWrite(EnablePin6, DMXSerial.read(startAddress +5));          
delay (500);
    }

      else {}

} // last loop
[/code]

**Comment by Gjorgjevikj** [comment]
Maybe I am not the right one to respond to this post, but as far as I can see the DMXSerial library explicitly specifies that it supports exact Arduino boards [Supported Boards and processors](https://github.com/mathertel/DMXSerial#supported-boards-and-processors) and LGT8FX... is not among them. Also if you look at the source of the DMXSerial library [DMXSerial_avr.h](https://github.com/mathertel/DMXSerial/blob/master/src/DMXSerial_avr.h) you can see that it performs some low level AVR register manipulation. LGT8FX is different (at lower level) and does not have these registers.

**Reply by LaZsolt** [reply]
The LGT8Fx low level registers and interrupts which used by DMXSerial, 100% same compatibility than ATmega328. 
I can't find any major errors at first sight.

**Reply by LaZsolt** [reply]
You must select 16 MHz clock source anyway. Otherwise it won't work.

**Reply by jg1uaa** [reply]
Is DMXSerial assumes 16MHz CPU clock? If LGT8F328P works with other frequency, I think this line in DMXSerial_avr.h needs to be modified...

https://github.com/mathertel/DMXSerial/blob/master/src/DMXSerial_avr.h#L147

```
// calculate prescaler from baud rate and cpu clock rate at compile time.
// This is a processor specific formula from the datasheet.
// It implements rounding of ((clock / 16) / baud) - 1.
#define CalcPreScale(B) (((((F_CPU) / 8) / (B)) - 1) / 2)
```

**Reply by jayzakk** [reply]
I could take a look on it tomorrow but I guess the dmxserial code needs to get fixed/modified.
Have dmx equipment, rs485 modules and oscilloscope at hand.

**Reply by jayzakk** [reply]
I checked it out and works out of the box with some cheap Showlite DMX desk.
Using the example DmxSerialRecv.ino without changes (except the LED pins) at default 32MHz. You must have done something wrong. Did you take care of routing D2 to RE+DE?

**Comment by ghost** [comment]
Many thanks.I gone make the test with 16Mhz.Keep you informed.


Verzonden via Yahoo Mail op Android 
 
  Op di, dec. 20, 2022 om 22:48 schreef ***@***.***>:   


You must select 16 MHz clock source anyway. Otherwise it won't work.

**Comment by ghost** [comment]
Thanks

Verzonden via Yahoo Mail op Android 
 
  Op wo, dec. 21, 2022 om 0:03 schreef ***@***.***>:   


I could take a look on it tomorrow but I guess the dmxserial code needs to get fixed/modified.
Have dmx equipment, rs485 modules and oscilloscope at hand.

**Comment by ghost** [comment]
If you want, I can also send you the code for the Arduino mega who is transmitting the dmx code.

Verzonden via Yahoo Mail op Android 
 
  Op wo, dec. 21, 2022 om 0:03 schreef ***@***.***>:   


I could take a look on it tomorrow but I guess the dmxserial code needs to get fixed/modified.
Have dmx equipment, rs485 modules and oscilloscope at hand.

**Comment by ghost** [comment]
Board settings in arduino ID 👍 
board : LGT8F328
Clock source : internal
Clock 16 MHz
Variant : 328P -SSOP20
Arduino has ISP : default (64)
Programmer : ArduinoISP

**Reply by LaZsolt** [reply]
@tandy1 

> Variant : 328P -SSOP20
Version SSOP20 not fit your project, because of not enough I/O pins.
Reason is: Digital I/O 10 and 11 are connected on the same pin.
```
const int EnablePin5 = 10;
const int EnablePin6 = 11;

```
For the DIP switches available pins are: 12, 13, A0, A1, A4, A5

So this package have three less I/O pins than you need.

Watch this picture:
https://blog.eletrogate.com/wp-content/webpc-passthru.php?src=https://blog.eletrogate.com/wp-content/uploads/2021/06/pinout-1.png&nocache=1

**Reply by LaZsolt** [reply]
Edit:

For the DIP switches available pins are: 2, 7, 12, 13, A0, A1, A4, A5

So this package **have one less** I/O pins than you need.

**Reply by jayzakk** [reply]
@LaZsolt , that variant selection itself is the problem - he wrote "LQFP32 PRO MINI NANO3" in thread start, so some would assume that the variant is set correctly, too ^^
Even the tested example code won't work anymore with SSOP20 setting.

@tandy1 : please use the correct variant "LQFP32", when your board is such.

**Reply by LaZsolt** [reply]
>  he wrote "LQFP32 PRO MINI NANO3" in thread start

I wasn't paying enough attention.

**Reply by ghost** [reply]
This is the board I have. 
https://nl.aliexpress.com/item/1005004582519143.html?spm=a2g0o.order_list.order_list_main.50.475f79d23IUVPV&gatewayAdapt=glo2nld

It has 8 analoge input pins and 12 digital I/O's.

**Reply by ghost** [reply]
LaZsolt ;  I changed the variant to LQFP32 but no result.

**Comment by jayzakk** [comment]
> The controller is based on this project I found on instructables.com. https://www.instructables.com/Build-Your-Own-DMX-Fixture-Arduino/

Side note: 

This schematics of that project are... uhm... let's say "wrong". Arduino D2 has to be connected to MAX485 Pins /RE (low enables receiver) and DE (high enables sender). The Library uses this pin to switch RX/TX, and ONLY using this way works when you need RX/TX to program the flash (via USB). D2 after reset is floating, so MAX485 does no connect its output to RX. If not, it's mixing up DMX RX and USB RX serial signals.

But problem then is: it drives the DMX bus HIGH for a around 1 second while resetting the arduino (as DE is pullup high), then competing with the DMX sender. This again is a flaw in the DMXSerial library, which should use 2 different I/O for driving the /RE and DE signals.

If you only intend to RECEIVE data, want to still use USB serial for programming, and not killing your DMX bus while resetting the device you should connect DE to GND, and /RE to D2.

**Reply by LaZsolt** [reply]
MAX485 Drivers are short-circuit current limited and are protected against excessive power dissipation by thermal shutdown circuitry that places the driver outputs into a high-impedance state.
It is a normal behaviour on the RS485 bus when 2 device start to sending data at the same time. So they won't kill each other.

**Reply by jayzakk** [reply]
They won't kill each other, that's true. But the RS485 bus is blocked (A/B difference gets too small. At least this is what my oscilloscope tells me, with only the DMX desk and the breadboard connected, the difference voltage drops lower than +-200mV). Something nobody wants on a DMX chain. Results may differ by position in the chain, strength of sender, cables, existence of termination resistor.

**Reply by LaZsolt** [reply]
Yes, it is true.

**Reply by ghost** [reply]
"This schematics of that project are... uhm... let's say "wrong". Arduino D2 has to be connected to MAX485 Pins /RE (low enables receiver) and DE (high enables sender). The Library uses this pin to switch RX/TX, and ONLY using this way works when you need RX/TX to program the flash (via USB). D2 after reset is floating, so MAX485 does no connect its output to RX. If not, it's mixing up DMX RX and USB RX serial signals."
Connection of the RS485 is : RO & DE connected to the - , DI not connected, RO connected to RX of the LGT8F328 board. 
DMX controller is only receiving data.
When I do the test with an original DMX card, everything is working. DMX information is correct sended by the master on the rs485 bus.
![DMX512](https://user-images.githubusercontent.com/24798309/208969458-0977933f-3fc5-48d9-93a2-bc71475c0eaa.jpeg)

**Comment by ghost** [comment, marked as answer]
The controller is working after changing the variant to LQFP32
Settings now : 
board : LGT8F328
Clock source : internal
Clock 16 MHz
Variant :LQFP32 (miniEVB nano-style ans WAVGAT)
Arduino has ISP : default (64)
Programmer : ArduinoISP


Many thanks for the support to all of you
